examples/pytorch/object-detection/run_object_detection.py

```diff
@@ -0,0 +1,496 @@
+    image_id: str, category: List[int], area: List[float], bbox: List[Tuple[float]]
+) -> dict:
+    """Format one set of image annotations to the COCO format
+
+    Args:
+        image_id (str): image id. e.g. "0001"
+        category (List[int]): list of categories/class labels corresponding to provided bounding boxes
+        area (List[float]): list of corresponding areas to provided bounding boxes
+        bbox (List[Tuple[float]]): list of bounding boxes provided in COCO format
+            ([center_x, center_y, width, height] in absolute coordinates)
+
+    Returns:
+        dict: {
+            "image_id": image id,
+            "annotations": list of formatted annotations
+        }
+    """
+    annotations = []
+    for i in range(len(category)):
+        formatted_annotation = {
+            "image_id": image_id,
+            "category_id": category[i],
+            "iscrowd": 0,
+            "area": area[i],
+            "bbox": list(bbox[i]),
+        }
+        annotations.append(formatted_annotation)
+
+    return {
+        "image_id": image_id,
+        "annotations": annotations,
+    }
+
+
+def convert_bbox_yolo_to_pascal(boxes: torch.Tensor, image_size: Tuple[int, int]) -> torch.Tensor:
+    """
+    Convert bounding boxes from YOLO format (x_center, y_center, width, height) in range [0, 1]
+    to Pascal VOC format (x_min, y_min, x_max, y_max) in absolute coordinates.
+
+    Args:
+        boxes (torch.Tensor): Bounding boxes in YOLO format
+        image_size (Tuple[int, int]): Image size in format (height, width)
+
+    Returns:
+        torch.Tensor: Bounding boxes in Pascal VOC format (x_min, y_min, x_max, y_max)
+    """
+    # convert center to corners format
+    boxes = center_to_corners_format(boxes)
+
+    # convert to absolute coordinates
+    height, width = image_size
+    boxes = boxes * torch.tensor([[width, height, width, height]])
+
+    return boxes
+
+
+def augment_and_transform_batch(
+    examples: Mapping[str, Any], transform: A.Compose, image_processor: AutoImageProcessor
+) -> BatchFeature:
+    """Apply augmentations and format annotations in COCO format for object detection task"""
+
+    images = []
+    annotations = []
+    for image_id, image, objects in zip(examples["image_id"], examples["image"], examples["objects"]):
+        image = np.array(image.convert("RGB"))
+
+        # apply augmentations
+        output = transform(image=image, bboxes=objects["bbox"], category=objects["category"])
+        images.append(output["image"])
+
+        # format annotations in COCO format
+        formatted_annotations = format_image_annotations_as_coco(
+            image_id, output["category"], objects["area"], output["bboxes"]
+        )
+        annotations.append(formatted_annotations)
+
+    # Apply the image processor transformations: resizing, rescaling, normalization
+    result = image_processor(images=images, annotations=annotations, return_tensors="pt")
+
+    return result
```

Comment:
I'd recommend not using the image processor at all. They're very slow, in part because the resizing is done in pillow (for historical reasons). This means every image is converted to numpy -> PIL.Image.Image -> numpy. Instead, I'd just do all of the transformations within the library of choice (albumentations, torchvision etc.) and use the image processor for values like `size` if needed

Reply:
I suggest leaving `image_processor` for correct input formatting and turning off padding and resizing.

```python
    image_processor = AutoImageProcessor.from_pretrained(
        model_args.image_processor_name or model_args.model_name_or_path,
        # At this moment we recommend using external transform to pad and resize images.
        # It`s faster and yields much better results for object-detection models.
        do_pad=False,
        do_resize=False,
        # We will save image size parameter in config just for reference
        size={"longest_edge": data_args.image_square_size},
        **common_pretrained_args,
    )
```

For padding and resizing I suggest the following strategy:
1) Resize the largest size of an image to `image_square_size`
2) Pad image to `image_square_size x image_square_size`

This strategy yields much better results in terms of mAP and also almost removes batch dependency for evaluation.
Here are two models trained with both strategies and evaluated for batch sizes `8` and `1`.

<img width="1107" alt="Screenshot 2024-04-30 at 11 44 49" src="https://github.com/huggingface/transformers/assets/31920396/ed682dc9-0c1a-4dc5-9bbe-afce31b90cf3">

Conditional DETR fine-tuned with such padding archives `mAP@0.5..0.95 = 0.5414`, while on [papers with code](https://paperswithcode.com/sota/object-detection-on-cppe-5) top model [TridentNet](https://paperswithcode.com/paper/cppe-5-medical-personal-protective-equipment) achieve lower `mAP@0.5..0.95 = 0.529` (probably, this LB is outdated, but still can be used as a reference).

Please, let me know what you think about this. Is it worth adding this strategy to image processors too?

Reply:
Thanks for the detailed explanation and runs. OK, sounds good to me! 

Regarding adding this to the image processors, it's a bit tricky as we need to account for backwards compatibility: even though this produces better results, DETR is a commonly used model and we shouldn't change the default behaviour. One option would be to add a flag to the image processor, which allows the user to pick the padding strategy, falling back to the current one by default

Reply:
Ok, got it. I will update accelerate example and make sure the tests pass.

Regarding image processors I understand the backward compatibility issues, here are some options we can implement

1) Add `preserve_image_ratio` flag (default is False). In combination with `size = {height: ...; width: ...}` it can be used to implement the suggested strategy. It is flexible, so we can use it even for non-square sizes. An image will be resized to respect height/width depending on the longest side and then padded to the specified size. But I dont like that it is not evident, and `preserve_image_ratio=False` may confuse for `size = {longest_edge: ...; shortest_edge: ...}` option, because this size option preserve image ratio by default.
2) As you suggested, add a flag `pad_strategy="batch"`. For `batch` it will follow current behaviour, for `pad_strategy="size"` it will pad to `size = {height: ...; width: ...}`. The problem with `size = {longest_edge: ...; shortest_edge: ...}` - we do not know which one is height and which is width, but we can raise an error for`pad_strategy="size"` to provide a correct size dict with height and width.

Both of them will require changing current resize and pad logic, but the second option seems better to me, let me know if you have any thoughts on that

Reply:
Agreed - let's go with the second option 🤝